### ISSUE TITLE
Bucket mounting should print friendlier error messages

### ISSUE THREAD

**Issue by concretevitamin** [skypilot-org/skypilot#1895]
Accidentally, I tried to mount to a path already taken up by the OS:
```
file_mounts:
  /tmp:
    name: tmp-workdir-zongheng
    source: ~/tmp-workdir
```
This throws a long error message:
```
<launching>
sky.exceptions.CommandError: Command (cat <<-\EOF > ~/.sky/mount_72018.sh
#!/usr/bin/env bash
set -e

MOUNT_PATH=/tmp
MOUNT_BINARY=gcsfuse

# Check if path is already mounted
if grep -q $MOUNT_PATH /proc/mounts ; then
    echo "Path already mounted - unmounting..."
    fusermount -uz "$MOUNT_PATH"
    echo "Successfully unmounted $MOUNT_PATH."
fi

# Install MOUNT_BINARY if not already installed
if [ -x "$(command -v gcsfuse)" ] && gcsfuse --version | grep -q 0.42.3; then
  echo "$MOUNT_BINARY already installed. Proceeding..."
else
  echo "Installing $MOUNT_BINARY..."
  wget -nc https://github.com/GoogleCloudPlatform/gcsfuse/releases/download/v0.42.3/gcsfuse_0.42.3_amd64.deb -O /tmp/gcsfuse.deb && sudo dpkg --install /tmp/gcsfuse.deb
fi

# Check if mount path exists
if [ ! -d "$MOUNT_PATH" ]; then
  echo "Mount path $MOUNT_PATH does not exist. Creating..."
  sudo mkdir -p $MOUNT_PATH
  sudo chmod 777 $MOUNT_PATH
else
  # Check if mount path contains files
  if [ "$(ls -A $MOUNT_PATH)" ]; then
    echo "Mount path $MOUNT_PATH is not empty. Please make sure its empty."
    exit 1
  fi
fi
echo "Mounting $SOURCE_BUCKET to $MOUNT_PATH with $MOUNT_BINARY..."
gcsfuse -o allow_other --implicit-dirs --stat-cache-capacity 4096 --stat-cache-ttl 5s --type-cache-ttl 5s --rename-dir-limit 10000 tmp-workdir-zongheng /tmp
echo "Mounting done."
) && chmod +x ~/.sky/mount_72018.sh && bash ~/.sky/mount_72018.sh && rm ~/.sky/mount_72018.sh failed with return code 1.
Failed to run command before rsync ~/tmp-workdir -> /tmp. Ensure that the network is stable, then retry.
```
We should

1. Skip printing the entire script
2. Print out the error messages from the underlying tool (probably something like `/tmp` is taken up)

**Comment by landscapepainter** [comment]
I can write a function to validate the destination in `Storage`` __init__`. What I can do to check if the destination is already taken by OS? Is it good enough to check if the owner of the directory is 'root' or not? @concretevitamin @romilbhardwaj

**Comment by concretevitamin** [comment]
If we can first investigate if the mounting script properly prints errors, that would be ideal. If so, the work for us would be just to surface whatever errors
```
bash ~/.sky/mount_72018.sh
```
prints. Perhaps by changing it to something like
```
... bash ~/.sky/mount_72018.sh || echo Mounting failed, detailed error: $<captured stderr> ...
```

**Comment by landscapepainter** [comment]
I did some testing and discovered that the error @concretevitamin encountered is due to mounted directory being non-empty. 

The testing includes the following:
1. Created a directory, `/doyoung-test`, in the remote VM with `sudo`, so that the directory's owner is `root`. I had it empty and tried to mount a Storage through `sky` and it went through smoothly. 
2. Created a directory in the remote VM without `sudo`, so that the directory's owner is the `gcpuser`. I created some text files in the directory and tried to mount a Storage through 'sky' and it has shown error message `Mount path /home/gcpuser/doyoung-test-1 is not empty. Please make sure its empty.` which is followed by the long message you mentioned above.
3. Mounted a Storage through `sky` to a directory that is taken up by OS (`root` owner and non-empty, `/sys`) while launching a remote VM with `sky launch`, and it has shown error message `Mount path /sys is not empty. Please make sure its empty.` which is followed by the long message you mentioned above.

The error message is long since the current `exception.py/CommandError.__init__()` prints out the entire bash script constructed in `mounting_utils.py/get_mounting_command()`. The current implementation does echo that the error is due to non-empty directory being attempted to be mounted (`Mount path /sys is not empty. Please make sure its empty.`) But it's easy to get lost due to the long bash script and other stdout `exception.py/CommandError.__init__()` echos.
 
However, unless we change how the error handling is done in `subprocess_utils.py/handle_returncode()`, it seems impossible to avoid printing the bash script and the long message when the mount path is non-empty. So I think it makes more sense to validate the mount path outside of the bash script before `store.mount_command(dst)` is ran in `cloud_vm_ray_backend.py/_execute_storage_mounts()` if we want to avoid `CommandError.__init__()` printing out the entire bash script.

I'll remove the following from `mounting_utils.py/get_mounting_command()`
```
          # Check if mount path contains files
          if [ "$(ls -A $MOUNT_PATH)" ]; then
            echo "Mount path $MOUNT_PATH is not empty. Please make sure its empty."
            exit 1
          fi
```
and do the mount path validation outside of the bash script in `cloud_vm_ray_backend.py/_execute_storage_mounts()`. What do you think? @romilbhardwaj @concretevitamin

**Comment by romilbhardwaj** [comment]
The mounting script can fail due to various reasons, including non-empty mount destination as mentioned above. Moving non-empty checks would solve this specific case, but if the script fails due to some other reason, we'll still get a long trace.

Can we instead fix `handle_returncode` or do special handling of CommandError when we run the mounting script to clean up the error message? Another option is to have the mounting commands in a file that is generated dynamically (instead of being a heredoc).

**Comment by romilbhardwaj** [comment]
> The mounting script can fail due to various reasons, including non-empty mount destination as mentioned above. Moving non-empty checks would solve this specific case, but if the script fails due to some other reason, we'll still get a long trace.
> 
> Can we instead fix handle_returncode or do special handling of CommandError when we run the mounting script to clean up the error message? Another option is to have the mounting commands in a file that is generated dynamically (instead of being a heredoc).

Ran into this today when fixing #3215. An error elsewhere in the heredoc (E.g., in the mount_cmd portion, due to permission issues) does not get surfaced. The return code returned by `SSHCommandRunner.run` is surprisingly 0 even though the same script when executed by ssh-ing into the instance returns an error code of 127.

**Comment by shethhriday29** [comment]
addressed by #3400 (currently in review process)
